### ISSUE TITLE
virtcontainers: add support for plugins

### ISSUE THREAD

**Issue by egernst** [kata-containers/runtime#51]
Add support for dynamically adding external virtcontainers interface implementations.  This allows for a clean way to allow for proprietary kata runtime extensions.

Suggest that we leverage golang's plugin [package](https://golang.org/pkg/plugin/).
Plugins to call into the virtcontainers plugin API (`RegisterPlugin()`)
Potentially supported plugin types:
* Hypervisors
* Metadata backend (For storing/fetching Kata runtime’s metadata)
* VM caching/cloning

Existing implementations (for example hypervisors) will be registered as plugins but still built-in, while external plugin implementations will be presented as dynamic libraries.

Let's use this issue to define the API.  This issue can be closed once the design is documented and plugin framework is merged.

**Comment by bergwolf** [comment]
ref: https://github.com/kata-containers/documentation/pull/27 on section [Plugin framework for external proprietary Kata runtime extensions](https://github.com/kata-containers/documentation/pull/27/files#diff-727ad799425e468e7ffa14abf330457cR81).

We might want to split it into three sub-issues one for each plugin types.
